### PR TITLE
Support ICAO equipment codes in flightplans

### LINKED ISSUE #124: Support ICAO equipment code in flight plan
Vatsim allows ICAO equipment codes now which are much more complex than the single-character FAA equipment codes. Andreas helpfully provided a good document explaining the codes ("Item 10").
[FPL-02.pdf](https://github.com/swift-project/pilotclient/files/7421332/FPL-02.pdf)
[FPL-03.pdf](https://github.com/swift-project/pilotclient/files/7421333/FPL-03.pdf)

### PR DESCRIPTION
**Pull request by ltoenning**
Fixes #124

With this PR, swift internally always uses ICAO equipment codes. If FAA codes are required (receiving or sending to a server that does not support them; this is controlled with a new FSD option), the ICAO codes are translated to FAA codes accordingly.

### CODE REVIEW

**Comment by ltoenning** [comment]
Since we received no issue reports from the testers, this is ready for merge.
@oktal3700 Would you like to review this PR first?

File: src/blackmisc/aviation/comnavequipment.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        CComNavEquipment() = default;
+
+        CComNavEquipment(ComNavEquipment comNavEquipment, CpdlcSatcomEquipment cpdlcSatcomEquipment);
+
+        explicit CComNavEquipment(QString equipment);
+
+        static QString flagToString(CpdlcSatcomEquipmentOption flag);
+        static QString flagToString(ComNavEquipmentOption flag);
+
+        static QStringList equipmentLetters();
```
Comment (oktal3700) [inline src/blackmisc/aviation/comnavequipment.h:79]: - [ ] Better name: `allEquipmentLetters()`

File: src/blackmisc/aviation/comnavequipment.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        bool hasEquipment(ComNavEquipmentOption equip) const { return m_equipment.testFlag(equip); }
+        bool hasEquipment(CpdlcSatcomEquipmentOption equip) const { return m_cpdlcSatcomEquipment.testFlag(equip); }
+
+        QStringList enabledOptions() const;
+
+        QString convertToQString(bool i18n = false) const;
```
Comment (oktal3700) [inline src/blackmisc/aviation/comnavequipment.h:86]: - [ ] Missing doxygen comments

File: src/blackmisc/aviation/flightplanaircraftinfo.cpp [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+    QString CFlightPlanAircraftInfo::convertToQString(bool) const
+    {
+        return asIcaoString();
+    }
+
+    void CFlightPlanAircraftInfo::parseIcaoEquipmentCode(const QString &equipment)
+    {
+        // Example: B789/H-SDE1E2E3FGHIJ2J3J4J5M1RWXY/LB1D1
+        QStringList firstSplit = equipment.split("/");
+        Q_ASSERT_X(firstSplit.size() == 3, Q_FUNC_INFO, "Cannot split string as required for the ICAO format");
```
Comment (oktal3700) [inline src/blackmisc/aviation/flightplanaircraftinfo.cpp:98]: - [ ] Is it correct that this should be an assert? Do we have a separate function for validating input?

File: src/blackmisc/aviation/flightplanaircraftinfo.cpp [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        {
+            m_comNavEquipment = CComNavEquipment();
+        }
+    }
+
+    void CFlightPlanAircraftInfo::parseFaaEquipmentCode(const QString &equipment)
+    {
+        // Example: H/A346/L
+        QStringList split = equipment.split('/');
+        Q_ASSERT_X(split.size() == 2 || split.size() == 3, Q_FUNC_INFO, "Cannot split string as required for the FAA format");
```
Comment (oktal3700) [inline src/blackmisc/aviation/flightplanaircraftinfo.cpp:153]: - [ ] Is it correct that this should be an assert? Do we have a separate function for validating input?

File: src/blackmisc/aviation/flightplanaircraftinfo.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+#include "blackmisc/aviation/ssrequipment.h"
+#include "blackmisc/aviation/waketurbulencecategory.h"
+
+#include <QString>
+
+BLACK_DECLARE_VALUEOBJECT_MIXINS(BlackMisc::Aviation, CFlightPlanAircraftInfo)
+
+namespace BlackMisc::Aviation
+{
+    class BLACKMISC_EXPORT CFlightPlanAircraftInfo : public CValueObject<CFlightPlanAircraftInfo>
```
Comment (oktal3700) [inline src/blackmisc/aviation/flightplanaircraftinfo.h:20]: - [ ] Missing doxygen comment

File: src/blackmisc/aviation/flightplanaircraftinfo.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    {
+    public:
+        CFlightPlanAircraftInfo() = default;
+
+        CFlightPlanAircraftInfo(const CAircraftIcaoCode &aircraftIcao, const CComNavEquipment &comNavEquipment,
```
Comment (oktal3700) [inline src/blackmisc/aviation/flightplanaircraftinfo.h:25]: - [ ] Missing doxygen comment

File: src/blackmisc/aviation/ssrequipment.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+BLACK_DECLARE_VALUEOBJECT_MIXINS(BlackMisc::Aviation, CSsrEquipment)
+
+namespace BlackMisc::Aviation
+{
+    /*! ICAO flightplan field 10b */
+    class BLACKMISC_EXPORT CSsrEquipment : public BlackMisc::CValueObject<CSsrEquipment>
+    {
+    public:
+        enum SSrEquipmentOption : int
```
Comment (oktal3700) [inline src/blackmisc/aviation/ssrequipment.h:18]: - [ ] Capitalization typo in `SSr` vs `Ssr`

File: src/blackmisc/aviation/ssrequipment.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+#include "blackmisc/blackmiscexport.h"
+#include "blackmisc/valueobject.h"
+
+BLACK_DECLARE_VALUEOBJECT_MIXINS(BlackMisc::Aviation, CSsrEquipment)
+
+namespace BlackMisc::Aviation
+{
+    /*! ICAO flightplan field 10b */
+    class BLACKMISC_EXPORT CSsrEquipment : public BlackMisc::CValueObject<CSsrEquipment>
```
Comment (oktal3700) [inline src/blackmisc/aviation/ssrequipment.h:15]: - [ ] Missing doxygen comments

File: src/blackmisc/aviation/ssrequipment.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            AdsBV2 = (1 << 15),
+            AdsCD1 = (1 << 16),
+            AdsCG1 = (1 << 17)
+        };
+
+        Q_DECLARE_FLAGS(SSrEquipment, SSrEquipmentOption);
+
+        CSsrEquipment() = default;
+
+        explicit CSsrEquipment(SSrEquipment equipment);
```
Comment (oktal3700) [inline src/blackmisc/aviation/ssrequipment.h:44]: - [ ] I think this constructor shouldn't be `explicit`

File: src/blackmisc/aviation/comnavequipment.cpp [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+        {
+            static const QString q("P2");
+            return q;
+        }
+        if (flag == SatvoiceRcp400)
+        {
+            static const QString q("P3");
+            return q;
+        }
+        return {};
```
Comment (oktal3700) [inline src/blackmisc/aviation/comnavequipment.cpp:218]:
```cpp
switch (flag)
{
case CpdlcAtn: return QStringLiteral("J1");
case CpdlcFansHfdl: return QStringLiteral("J2");
case CpdlcFansVdlA: return QStringLiteral("J3");
//...
default: return {};
}
```

File: src/blackmisc/aviation/comnavequipment.cpp [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+        {
+            const QString q("Y");
+            return q;
+        }
+        if (flag == Other)
+        {
+            const QString q("Z");
+            return q;
+        }
+        return {};
```
Comment (oktal3700) [inline src/blackmisc/aviation/comnavequipment.cpp:343]:
```cpp
switch (flag)
{
case Standard: return QStringLiteral("S");
case Gbas: return QStringLiteral("A");
case Lpv: return QStringLiteral("B");
//...
default: return {};
}
```

File: src/blackgui/components/flightplancomponent.cpp [hunk 12]
```diff
@@ -70,6 +72,11 @@ namespace BlackGui::Components
         ui->setupUi(this);
         this->setCurrentIndex(0);
 
+        // Initialize wake turbulence category selection
+        std::for_each(m_wakeTurbulenceCategories.cbegin(), m_wakeTurbulenceCategories.cend(), [this](const WakeTurbulenceEntry &item) {
+            ui->cb_Wtc->addItem(item.m_name);
+        });
```
Comment (oktal3700) [inline src/blackgui/components/flightplancomponent.cpp:78]:
```cpp
for (const WakeTurbulenceEntry &item : std::as_const(m_wakeTurbulenceCategories))
{
    ui->cb_Wtc->addItem(item.m_name);
}
```

File: src/blackgui/components/flightplancomponent.h [hunk 13]
```diff
@@ -264,6 +269,35 @@ namespace BlackGui::Components
...
+        //! Update the selection and label according to current NAV/COM equipment
+        void updateNavComEquipmentUi();
+
+        //! Update the selection and label according to current SSR equipment
+        void updateSsrEquipmentUi();
+
+        //! Update the wake turbulence category selector
+        void updateWakeTurbulenceCategorySelector(const BlackMisc::Aviation::CWakeTurbulenceCategory &wtc);
+
+        BlackMisc::Aviation::CWakeTurbulenceCategory getSelectedWakeTurbulenceCategory() const;
```
Comment (oktal3700) [inline src/blackgui/components/flightplancomponent.h:293]: - [ ] Missing doxygen comment

File: src/blackmisc/aviation/flightplanaircraftinfo.cpp (hunk 4 above)
Comment (ltoenning) [reply src/blackmisc/aviation/flightplanaircraftinfo.cpp:153]: This is checked before calling the method here:

https://github.com/swift-project/pilotclient/blob/5fc492bc8d576e66f92a24e631c072a340a1dd7a/src/blackmisc/aviation/flightplanaircraftinfo.cpp#L25-L28

And as the method is ``private``, it is only used internally.

**Review by oktal3700 — APPROVED** [review]

**Pull request merged**